feat(Data/Fin/Basic): Refactor succAbove and predAbove.

linesthatinterlace opened the pull request:
We currently implemented `succAbove` and `predAbove` in a way that encourages users to reduce things down to fiddling linear arithmetic, which makes it difficult to prove complicated identities.

---
It is possible to re-define these and improve the proofs dramatically. An aid to this is to add "predAboveOfNe", which broadly is similar to `predAbove` but is a true inverse to `succAbove` in the appropriate manner. One can then prove results about this which let the simplifier make short work of identities that would otherwise be fairly complicated.

This also has the advantage of making insertNth easier to use. (To be demonstrated!)

<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->
[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

kbuzzard commented on Mathlib/AlgebraicTopology/SimplexCategory.lean:
```diff
@@ -212,23 +212,18 @@ def σ {n} (i : Fin (n + 1)) : ([n + 1] : SimplexCategory) ⟶ [n] :=
-  · simpa only [Fin.le_iff_val_le_val, ← Nat.lt_succ_iff, Nat.succ_eq_add_one, ← Fin.val_succ,
-      j.succ_pred, Fin.lt_iff_val_lt_val] using H
+  rw [← δ_comp_δ, Fin.succ_pred]
+  rw [← Fin.succ_le_succ_iff, Fin.succ_pred, ← Fin.castSucc_lt_iff_succ_le]
```
Does `rwa` close this in one line?

kbuzzard commented on Mathlib/Data/Fin/Basic.lean:
```diff
@@ -934,6 +964,46 @@ def castSuccEmb : Fin n ↪o Fin (n + 1) :=
 
+@[simp]
+theorem castSucc_le_castSucc_iff : castSucc a ≤ castSucc b ↔ a ≤ b := Iff.rfl
+@[simp]
```
```suggestion

@[simp]
```
and similarly for the next few simp lemmas

linesthatinterlace replied:
To be clear, this is just a spacing thing?

kbuzzard commented on Mathlib/Data/Fin/Basic.lean:
```diff
@@ -1089,8 +1162,125 @@ section Pred
 
+section CastPred
+
+/-- TODO -/
```
Delete? (if done?)

linesthatinterlace replied:
Needs a docstring writing I think is what this is saying.

kbuzzard commented on Mathlib/Data/Fin/Basic.lean:
```diff
@@ -1377,488 +1567,1212 @@ section SuccAbove
+theorem succAbove_of_castSucc_lt {p : Fin (n + 1)} {i : Fin n} (h : castSucc i < p) :
+    p.succAbove i = castSucc i := dif_pos h
+#align fin.succ_above_below Fin.succAbove_of_castSucc_lt
+theorem succAbove_of_succ_le {p : Fin (n + 1)} {i : Fin n} (h : succ i ≤ p) :
```
blank line before theorem?

linesthatinterlace replied:
See, Yael was telling me recently that they like to group similar theorems together.

alexkeizer replied:
Personally, when grouping things I like to it with a header comment like:
```lean
/-!
## Foo theorems
-/
```

kbuzzard commented on Mathlib/Data/Fin/Basic.lean:
```diff
@@ -1377,488 +1567,1212 @@ section SuccAbove
+    (hp : p ≤ r.castSucc) (hq : r.castSucc < q) :
+    (p.succAbove r).castSucc = (q.succAbove r).succ :=
+  succAbove_castSucc_eq_succAbove_succ_iff.mpr ⟨hp, hq⟩
+theorem succAbove_castSucc_eq_succ_of_lt_succ_succ_le {p q : Fin (n + 1)} {r : Fin n}
```
etc

kbuzzard reviewed (commented):
It's a shame that github makes such a mess of the diff for Data.Fin.Basic, making it much harder to review -- but the impact in every other file seems highly positive! Very nice!

linesthatinterlace commented:
It's easier to see the difference if you just review the total change of all the patches. But yeah I think the main changes in Data/Fin/Basic.lean kinda need a line by line review. I am especially worried I've stated lemmas in inappropriate generality somewhere, or failed to "mirror" some theorems between predAbove and succAbove (as a rule of thumb, every result one should correspond to one in the other, more or less - there's a fundamental symmetry here).

Note that the impact on other files is in some cases because the proofs are effectively in Basic.lean now! But also I've generally improved them, and part of the issue was in fact that things like the simplicial proofs really ought not to be only in the language of simplicial categories elsewhere, because they do genuinely occur in practice.

alexkeizer commented on Mathlib/Data/Fin/Basic.lean:
```diff
@@ -1089,8 +1162,125 @@ section Pred
+
+/-- TODO -/
+
+@[inline] def castPred (i : Fin (n + 1)) (h : i ≠ last _) : Fin n := ⟨i, val_lt_last h⟩
+#align fin.cast_pred Fin.castPred
```
From what I can tell, this is not just refactoring `castPred`, but changing it's meaning from mapping `last (n+1) => last n` to restricting the domain with the `i /= last _` assumption.
I don't think this is desirable from a backwards compatibility POV. Maybe we rename this definition to `castPredOfNe`, and define `castPred` in terms of it and `if-then-else`?

linesthatinterlace replied:
No - I'm making the change to make it consistent with pred. If pred was instead predOfNe I would agree, but the current state of affairs where they are inconsistent is not desirable, and I think this is better.

Note that the current castPred is exactly predAbove 0.

linesthatinterlace replied:
It's confusing that pred and castPred currently work differently, and the only reason for this that I can see is that someone did a so-so job in it in the first place.

linesthatinterlace commented on Mathlib/Data/Fin/Basic.lean:
```diff
@@ -510,6 +521,11 @@ theorem val_one'' {n : ℕ} : ((1 : Fin (n + 1)) : ℕ) = 1 % (n + 1) :=
   rfl
 #align fin.one_val Fin.val_one''
 
+theorem castSucc_one' [NeZero n] : castSucc (1 : Fin (n + 1)) = 1 := by
+  rw [ext_iff, coe_castSucc, val_one', val_one, mod_succ_eq_iff_lt, succ_eq_one_add,
+  lt_add_iff_pos_right]
+  exact NeZero.pos n
```
Note that castSucc_one' exists, with the NeZero assumption. The answer to this is "because Std doesn't have NeZero", I guess. Technically you could get rid of all the n+2 lemmas, but the convention seems to generally be to have both. It might be that's just a limitation of the fact that Std has a lot of these though.

It is never correct in my view to *only* have the n+2 case, you really want to avoid having to ever eliminate the trivial case, even though this can often be annoying. For instance, predAbove, because it takes in a Fin n as an argument, could really be defined for n+1 only - but you don't need this to be true, and if you can do it in generality, you should.

linesthatinterlace commented on Mathlib/Data/Fin/Basic.lean:
```diff
@@ -934,6 +964,46 @@ def castSuccEmb : Fin n ↪o Fin (n + 1) :=
+@[simp]
+theorem castSucc_le_castSucc_iff : castSucc a ≤ castSucc b ↔ a ≤ b := Iff.rfl
+@[simp]
+theorem succ_le_castSucc_iff : succ a ≤ castSucc b ↔ a < b :=
+  by rw [le_castSucc_iff, succ_lt_succ_iff]
```
Oh yes, my bad. We should add this to the style checker...

linesthatinterlace commented on Mathlib/Data/Fin/Basic.lean:
```diff
@@ -1089,8 +1162,125 @@ section Pred
+theorem pred_last (h := last_pos'.ne') :
+    pred (last (n + 1)) h = last n := by simp_rw [← succ_last, pred_succ]
+
+theorem pred_lt_iff {i : Fin (n + 1)} (hi : i ≠ 0) : pred i hi < j ↔ i < succ j := by
+  rw [← succ_lt_succ_iff, succ_pred]
+theorem lt_pred_iff {i : Fin (n + 1)} (hi : i ≠ 0) : j < pred i hi ↔ succ j < i := by
+  rw [← succ_lt_succ_iff, succ_pred]
+theorem pred_le_iff {i : Fin (n + 1)} (hi : i ≠ 0) : pred i hi ≤ j ↔ i ≤ succ j := by
+  rw [← succ_le_succ_iff, succ_pred]
+theorem le_pred_iff {i : Fin (n + 1)} (hi : i ≠ 0) : j ≤ pred i hi ↔ succ j ≤ i := by
+  rw [← succ_le_succ_iff, succ_pred]
```
I'm not certain there is one.

linesthatinterlace commented on Mathlib/Data/Fin/Basic.lean:
```diff
@@ -1377,488 +1567,1212 @@ section SuccAbove
 /-- `succAbove p i` embeds `Fin n` into `Fin (n + 1)` with a hole around `p`. -/
 def succAbove (p : Fin (n + 1)) (i : Fin n) : Fin (n + 1) :=
-  if i.1 < p.1 then castSucc i else succ i
+  if castSucc i < p then i.castSucc else i.succ
```
Essentially completely at random. Here it is because it's nice to have the " I < p" part clearly visible. But I could just remove all dot notation...

linesthatinterlace commented on Mathlib/Data/Fin/Basic.lean:
```diff
@@ -1089,8 +1162,125 @@ section Pred
+#align fin.coe_cast_pred Fin.coe_castPred
+
+@[simp]
+theorem castPred_castSucc {i : Fin n} (h' := (castSucc_lt_last i).ne) :
+    castPred (castSucc i) h' = i := rfl
```
Ye-es. Notice that it isn't a regular argument - it's an optParam. We generally do this for cases like this where you need the proof to state the statement, because it makes rewriting easier. This is fairly standard I think.

alexkeizer replied:
I didn't know `optParam`s make rewrites easier, and it indeed seems a more common pattern, so I'll retract my comment

linesthatinterlace commented on Mathlib/Data/Fin/Tuple/Basic.lean:
```diff
@@ -739,8 +739,9 @@ def succAboveCases {α : Fin (n + 1) → Sort u} (i : Fin (n + 1)) (x : α i)
-    if hlt : j < i then @Eq.recOn _ _ (fun x _ ↦ α x) _ (succAbove_castLT hlt) (p _)
-    else @Eq.recOn _ _ (fun x _ ↦ α x) _ (succAbove_pred <| (Ne.lt_or_lt hj).resolve_left hlt) (p _)
+    if hlt : j < i then @Eq.recOn _ _ (fun x _ ↦ α x) _ (succAbove_castPred_of_lt hlt) (p _)
+    else @Eq.recOn _ _ (fun x _ ↦ α x) _ (succAbove_pred_of_gt <|
+    (Ne.lt_or_lt hj).resolve_left hlt) (p _)
```
I'm fairly sure this isn't my addition, the definition is just like that. It would be good if it was better but one thing at a time.

alexkeizer replied:
Sure, let's close this comment, then! (It would be nice to have a way to view diffs modulo re-ordered definitions...)

linesthatinterlace commented on Mathlib/Algebra/BigOperators/Fin.lean:
```diff
@@ -285,16 +285,18 @@ theorem inv_partialProd_mul_eq_contractNth {G : Type*} [Group G] (g : Fin (n + 1
   rcases lt_trichotomy (k : ℕ) j with (h | h | h)
-  · rwa [succAbove_below, succAbove_below, partialProd_right_inv, contractNth_apply_of_lt]
+  · rwa [succAbove_of_castSucc_lt, succAbove_of_castSucc_lt, partialProd_right_inv,
+    contractNth_apply_of_lt]
     · assumption
     · rw [castSucc_lt_iff_succ_le, succ_le_succ_iff, le_iff_val_le_val]
       exact le_of_lt h
-  · rwa [succAbove_below, succAbove_above, partialProd_succ, castSucc_fin_succ, ← mul_assoc,
+  · rwa [succAbove_of_castSucc_lt, succAbove_of_le_castSucc, partialProd_succ, castSucc_fin_succ,
+      ← mul_assoc,
       partialProd_right_inv, contractNth_apply_of_eq]
```
Yeah I'm not actually sure. Maybe this is written down somewhere.

Parcly-Taxel replied:
Most people always _do_ indent multiline `rw` and you should do it too.

alexkeizer commented:
I've gone through part of the `Fin.Basic` file, but not all: the diff is rather big and hard to thoroughly review.
So far, though, I really like the proposed refactor, and would say it's a clear improvement to be able to reason more about operations on `Fin`s, rather than having to reason about the underlying `Nat` values!

linesthatinterlace commented:
This PR is going to be closed, and #10283 will replace its substantative results. (It is now stale, and I have extracted the key changes and implemented them piecemeal.)

joneugster commented:
The authour mentions they extracted the relevant aspects of this PR, so I assume this can be closed?

Generally, I think how to proceed here, or in equivalent PRs (if they are big enough) in the future is the following:

1. make one PR where just do the renaming you want to, add `@[deprecated (since := "2025-01-11")] alias ...` for each of them, and do all the renames where the deprecation warning appears.
2. make a second, dependend PR with just addition of any new lemmas, but don't use them yet
3. and a third PR with the refactors using these new lemmas

It is not necessary to make the full split into three PRs, but it helps reducing the review time drastically, hoping preventing it from going stale.

linesthatinterlace commented:
Yeah, close it. I've just been too busy with trying to get my thesis finished to be honest, I'm sorry for letting this go a bit stale.
